app/src/main/java/org/onionshare/android/ShareManager.kt

```diff
@@ -42,132 +40,125 @@ class ShareManager @Inject constructor(
-        if (shareState.value is ShareUiState.ErrorAddingFile) fileManager.resetError()
-        shouldStop.value = false
+        _shareState.value = ShareUiState.Starting(0, 0)
         // Attention: We'll launch sharing in Global scope, so it survives ViewModel death,
         // because this gets called implicitly by the ViewModel in ViewModelScope
         @Suppress("OPT_IN_USAGE")
         startSharingJob = GlobalScope.launch(Dispatchers.IO) {
-            coroutineScope {
+            coroutineScope mainScope@{
+                fun stopOnError(error: ShareUiState.Error) {
+                    notificationManager.onError()
+                    _shareState.value = error
+                    // stop in a new scope to not cause deadlock when waiting for startSharingJob to complete
+                    GlobalScope.launch {
+                        stopSharing(error)
+                    }
+                    this@mainScope.cancel()
+                }
                 // call ensureActive() before any heavy work to ensure we don't continue when cancelled
                 ensureActive()
                 // When the current scope gets cancelled, the async routine gets cancelled as well
                 val fileTask = async { fileManager.zipFiles() }
                 // start tor and onion service
-                val torTask = async { torManager.start() }
-                fileTask.await()
+                val torTask = async {
+                    try {
+                        torManager.start()
+                    } catch (e: Exception) {
+                        LOG.error("Error starting Tor: ", e)
+                        stopOnError(ShareUiState.ErrorStarting(errorMsg = e.toString()))
+                    }
+                }
+                // wait for tor.start() to return before starting to observe, actual startup happens async
                 torTask.await()
+                LOG.info("Tor task returned")
+                // start progress observer task
+                val observerTask = async {
+                    fileManager.zipState.combine(torManager.state) { zipState, torState ->
+                        onStarting(zipState, torState)
+                    }.transformWhile { shareUiState ->
+                        emit(shareUiState)
+                        // only continue collecting while we are starting (otherwise would never stop collecting)
+                        shareUiState is ShareUiState.Starting
+                    }.collect { shareUiState ->
+                        LOG.info("New share state: $shareUiState")
+                        _shareState.value = shareUiState
+                        if (shareUiState is ShareUiState.Error) stopOnError(shareUiState)
+                    }
+                    LOG.info("Observer task finished.")
+                }
+                ensureActive()
+                val zipResult = fileTask.await()
+                if (zipResult is ZipResult.Zipped) {
```

Comment:
Kotlin noob question: would it be worthwhile and/or idiomatic to use `when` here so we get an error if we later add a third ZipResult that's not handled here?

Reply STOP to unsubscribe from Kotlin noob questions

Reply:
yes, good suggestion. I think you are getting ready to write kotlin code yourself :P